AwsEncryptionSDK/runtimes/net/Examples/Keyring/AwsKmsHierarchical/AwsKmsHierarchicalKeyringBranchKeySupplier.cs

```diff
@@ -0,0 +1,227 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+using Amazon.DynamoDBv2;
+using Amazon.KeyManagementService;
+using AWS.Cryptography.EncryptionSDK;
+using AWS.Cryptography.KeyStore;
+using AWS.Cryptography.MaterialProviders;
+using Xunit;
+
+/// <summary>
+/// This example sets up the Hierarchical Keyring, 
+/// which establishes a key hierarchy
+/// where "branch" keys are persisted in DynamoDb.
+/// These branch keys are used to protect your data keys,
+/// and these branch keys are themselves protected by a KMS Key.
+/// 
+/// Establishing a key hierarchy like this has two benefits:
+/// 
+/// First, by caching the branch key material, and only calling
+/// KMS to re-establish authentication regularly according to your configured TTL,
+/// you limit how often you need to call KMS to protect your data.
+/// This is a performance/security tradeoff, where your authentication, audit, and
+/// logging from KMS is no longer one-to-one with every encrypt or decrypt call.
+/// Additionally KMS Cloudtrail cannot be used to distinguish Encrypt and Decrypt calls,
+/// and you cannot restrict who has Encryption rights from who has Decryption rights
+/// since they both ONLY need KMS:Decrypt.
+/// However, the benefit is that you no longer have to make a
+/// network call to KMS for every encrypt or decrypt.
+/// 
+/// Second, this key hierarchy facilitates cryptographic isolation
+/// of a tenant's data in a multi-tenant data store.
+/// Each tenant can have a unique Branch Key,
+/// that is only used to protect the tenant's data.
+/// You can either statically configure a single branch key
+/// to ensure you are restricting access to a single tenant,
+/// or you can implement an interface that selects 
+/// the Branch Key based on the Encryption Context.
+/// 
+/// This example demonstrates configuring a Hierarchical Keyring
+/// with a Branch Key ID Supplier to encrypt and decrypt data for
+/// two separate tenants.
+/// 
+/// This example requires access to the DDB Table where you are storing the Branch Keys.
+/// This table must be configured with the following
+/// primary key configuration:
+/// - Partition key is named "partition_key" with type (S)
+/// - Sort key is named "sort_key" with type (S)
```

Comment:
Suggestion: 
If we are not going to create or version branch keys,
we do not need write permission.

```suggestion
/// This example requires read access to the DDB Table
/// where you are storing the Branch Keys.
/// primary key configuration:
/// - Partition key is named "partition_key" with type (S)
/// - Sort key is named "sort_key" with type (S)
```

Reply:
Although the example won't create or version, someone reading this example and trying to run it won't know what permissions they need to successfully run the example

Reply:
> someone reading this example and trying to run it won't know what permissions they need to successfully run the example

Yes. Hence my suggestion that we document it.  
In meetings, we have celebrated the Permission Boundary of the H-Keyring
and Key Store.

The Key Store implicitly supports two levels of Actors:
- Keyring Users : Read the Table, Decrypt from KMS
- Key Store Admins : Read and Write to the table; Generate, ReEncrypt, Decrypt from KMS

I do not want to bother documenting all of that here.
We could, and we SHOULD, one day, document all of this.

For now, we can imply this Actor distinction to our customers in this example.

After all, 
this Example does NOT include any Branch Key Creation or Version calls,
commented out or otherwise.